Crash with struct field `?time.Time`

### Describe the bug

(See MCE)

### Expected Behavior

```
Foo{
    t: Option(time.Time{})
}
```

### Current Behavior

```
/tmp/v_501/main.14891750470203069869.tmp.c:2446:30: error: passing '_option_time__Time' (aka 'struct _option_time__Time') to parameter of incompatible type 'time__Time' (aka 'struct time__Time')
        string _t1 = time__Time_str(it.t);
                                    ^~~~
/tmp/v_501/main.14891750470203069869.tmp.c:2104:34: note: passing argument to parameter 't' here
string time__Time_str(time__Time t);
                                 ^
/tmp/v_501/main.14891750470203069869.tmp.c:2591:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
1 warning and 1 error generated.
builder error:
==================
C error. This should never happen.
```

### Reproduction Steps

## MCE follows

```v
import time

struct Foo {
  t? time.Time
}

fn main() {
  println(Foo{})
}
```

`v -run main.v` or `v -cg run main.v`


### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 e066d1d

### Environment details (OS name and version, etc.)

```
OS: macos, macOS, 13.2.1, 22D68
Processor: 10 cpus, 64bit, little endian, Apple M1 Max
CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)

getwd: {{HOME}}/Projects/v
vmodules: {{HOME}}/.vmodules
vroot: {{HOME}}/.homebrew/Cellar/vlang/HEAD-e066d1d/libexec
vexe: {{HOME}}/.homebrew/Cellar/vlang/HEAD-e066d1d/libexec/v
vexe mtime: 2023-02-17 02:56:58
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 e066d1d

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: 4.0.1-24-ge3580359 (49047 commit(s) behind V master)
.git/config present: false
thirdparty/tcc status: thirdparty-macos-arm64 173c526e
```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -899,7 +899,7 @@ fn (mut g Gen) gen_str_for_struct(info ast.Struct, styp string, typ_str string, 
 		mut field_styp := sftyp.replace('*', '')
 		field_styp_fn_name := if sym_has_str_method {
 			mut field_fn_name := if ftyp_noshared.has_flag(.option) {
-				'${field_styp}_str'
+				g.get_str_fn(ftyp_noshared)
 			} else {
 				left_cc_type := g.cc_type(ftyp_noshared, false)
 				left_fn_name := util.no_dots(left_cc_type)

--- a/vlib/x/json2/README.md
+++ b/vlib/x/json2/README.md
@@ -15,7 +15,7 @@ mut:
 	name     string
 	age      ?int = 20
 	birthday time.Time
-	deathday ?time.Time = none
+	deathday ?time.Time
 }
 
 fn main() {
@@ -39,11 +39,11 @@ mut:
 	name     string
 	age      ?int = 20
 	birthday time.Time
-	deathday ?time.Time = none
+	deathday ?time.Time
 }
 
 fn main() {
-	resp := '{"name": "Bob", "age": 20, "birthday": ${time.now()}}'
+	resp := '{"name": "Bob", "age": 20, "birthday": "${time.now()}"}'
 	person := json2.decode[Person](resp)!
 	/*
 	struct Person {

--- a/vlib/x/json2/json2.v
+++ b/vlib/x/json2/json2.v
@@ -52,12 +52,12 @@ pub fn decode[T](src string) !T {
 		} $else $if field.typ is i16 {
 			typ.$(field.name) = res[json_name]!.int()
 		} $else $if field.typ is i32 {
-			// typ.$(field.name) = res[field.name]!.i32()
+			typ.$(field.name) = i32(res[field.name]!.int())
 		} $else $if field.typ is i64 {
 			typ.$(field.name) = res[json_name]!.i64()
 		} $else $if field.typ in [?u8, ?i8, ?i16, ?i32, ?u16, ?u32, ?u64, ?int] {
 			if json_name in res {
-				typ.$(field.name) = res[json_name]!.int()
+				typ.$(field.name) = res[json_name]!.i64()
 			}
 		} $else $if field.typ is f32 {
 			typ.$(field.name) = res[json_name]!.f32()

--- a/vlib/x/json2/json2.v
+++ b/vlib/x/json2/json2.v
@@ -55,16 +55,40 @@ pub fn decode[T](src string) !T {
 			// typ.$(field.name) = res[field.name]!.i32()
 		} $else $if field.typ is i64 {
 			typ.$(field.name) = res[json_name]!.i64()
+		} $else $if field.typ in [?u8, ?i8, ?i16, ?i32, ?u16, ?u32, ?u64, ?int] {
+			if json_name in res {
+				typ.$(field.name) = res[json_name]!.int()
+			}
 		} $else $if field.typ is f32 {
 			typ.$(field.name) = res[json_name]!.f32()
+		} $else $if field.typ is ?f32 {
+			if json_name in res {
+				typ.$(field.name) = res[json_name]!.f32()
+			}
 		} $else $if field.typ is f64 {
 			typ.$(field.name) = res[json_name]!.f64()
+		} $else $if field.typ is ?f64 {
+			if json_name in res {
+				typ.$(field.name) = res[json_name]!.f64()
+			}
 		} $else $if field.typ is bool {
 			typ.$(field.name) = res[json_name]!.bool()
+		} $else $if field.typ is ?bool {
+			if json_name in res {
+				typ.$(field.name) = res[json_name]!.bool()
+			}
 		} $else $if field.typ is string {
 			typ.$(field.name) = res[json_name]!.str()
+		} $else $if field.typ is ?string {
+			if json_name in res {
+				typ.$(field.name) = res[json_name]!.str()
+			}
 		} $else $if field.typ is time.Time {
 			typ.$(field.name) = res[field.name]!.to_time()!
+		} $else $if field.typ is ?time.Time {
+			if json_name in res {
+				typ.$(field.name) = res[field.name]!.to_time()!
+			}
 		} $else $if field.is_array {
 			// typ.$(field.name) = res[field.name]!.arr()
 		} $else $if field.is_struct {

--- a/vlib/x/json2/json2.v
+++ b/vlib/x/json2/json2.v
@@ -55,9 +55,41 @@ pub fn decode[T](src string) !T {
 			typ.$(field.name) = i32(res[field.name]!.int())
 		} $else $if field.typ is i64 {
 			typ.$(field.name) = res[json_name]!.i64()
-		} $else $if field.typ in [?u8, ?i8, ?i16, ?i32, ?u16, ?u32, ?u64, ?int] {
+		} $else $if field.typ is ?u8 {
 			if json_name in res {
-				typ.$(field.name) = res[json_name]!.i64()
+				typ.$(field.name) = ?u8(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?i8 {
+			if json_name in res {
+				typ.$(field.name) = ?i8(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?u16 {
+			if json_name in res {
+				typ.$(field.name) = ?u16(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?i16 {
+			if json_name in res {
+				typ.$(field.name) = ?i16(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?u32 {
+			if json_name in res {
+				typ.$(field.name) = ?u32(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?i32 {
+			if json_name in res {
+				typ.$(field.name) = ?i32(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?u64 {
+			if json_name in res {
+				typ.$(field.name) = ?u64(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?i64 {
+			if json_name in res {
+				typ.$(field.name) = ?i64(res[json_name]!.i64())
+			}
+		} $else $if field.typ is ?int {
+			if json_name in res {
+				typ.$(field.name) = ?int(res[json_name]!.i64())
 			}
 		} $else $if field.typ is f32 {
 			typ.$(field.name) = res[json_name]!.f32()

